src/Hash.mo

```diff
@@ -22,8 +22,10 @@ module {
     ha == hb
   };
 
-  public func hash(i : Nat) : Hash {
-    let j = Prim.natToNat32(i);
+  /// Computes a hash from the least significant 32-bits of `n`, ignoring other bits.
+  /// @deprecated For large `Nat` values consider using a bespoke hash function that considers all of the argument's bits.
```

Comment:
Can't we just fix the hash function instead?

Reply:
That would be sensible - LibTomMath also provides one but I don't know how suitable it is and we have to also make sure it agrees with any value returned for compact representations.
I propose we do that and then add the functions to Nat and Int modules, and deprecate/remove this broken one for good.

Reply:
This is just a quick fix to make the function total at least, even if it's a terrible hash function.

Reply:
Can we use this: https://github.com/libtom/libtommath/blob/5108f12350b6daa4aa5dbc846517ad1db2f8388a/tommath.h#L498?

Reply:
Probably, unless we need to keep the hash stable across motoko-base versions?

Reply:
https://github.com/libtom/libtommath/blob/5108f12350b6daa4aa5dbc846517ad1db2f8388a/doc/bn.tex#L1459 is concerning in that regard

Reply:
I don't think we plan to change that parameter.

Reply:
We probably want to keep the hash stable. Otherwise, an upgrade in the base may cause data corruption in `HashMap`.

One hacky fix is to `debug_print` the number and use `Text.hash`.

Reply:
That is a good reason. In that case, we should make this official: document this guarantee, and have enough test cases checking the concrete value for a few inputs, so that we'd notice if someone tries to change the hash.

Reply:
`HashMap` is fine, but `Trie` can be a `stable var` which depends on both Hash and Eq.

Reply:
Right this is why I suggest we keep the current function in Hash (for stability), but make it total not partial - or is that dangerous?

I'd like to add better Nat/Int hash functions later, and put them in the Nat/Int modules to distinguish them from this legacy one.